manifest-updater/manifest/storage.go

```diff
@@ -0,0 +1,260 @@
+package manifest
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"cloud.google.com/go/storage"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/rs/zerolog/log"
+)
+
+type Storage interface {
+	Writer
+	Fetcher
+}
+
+// Writer writes manifests to some storage
+type Writer interface {
+	// WriteDataShareProcessorSpecificManifest writes the provided manifest for
+	// the provided share processor name in the writer's backing storage, or
+	// returns an error on failure.
+	WriteDataShareProcessorSpecificManifest(manifest DataShareProcessorSpecificManifest, dataShareProcessorName string) error
+	// WriteIngestorGlobalManifest writes the provided manifest to the writer's
+	// backing storage, or returns an error on failure.
+	WriteIngestorGlobalManifest(manifest IngestorGlobalManifest) error
+}
+
+// Fetcher fetches manifests from some storage
+type Fetcher interface {
+	// FetchDataShareProcessorSpecificManifest fetches the specific manifest for
+	// the specified data share processor and returns it, if it exists and is
+	// well-formed. Returns (nil,  nil) if the  manifest does not exist.
+	// Returns (nil, error) if something went wrong while trying to fetch or
+	// parse the manifest.
+	FetchDataShareProcessorSpecificManifest(dataShareProcessorName string) (*DataShareProcessorSpecificManifest, error)
+	// IngestorGlobalManifestExists returns true if the global manifest exists
+	// and is well-formed. Returns (false, nil) if it does not exist. Returns
+	// (false, error) if something went wrong while trying to fetch or parse the
+	// manifest.
+	IngestorGlobalManifestExists() (bool, error)
+}
+
+// Bucket specifies the cloud storage bucket where manifests are stored
+type Bucket struct {
+	// Bucket is the name of the bucket, without any URL scheme
+	Bucket string `json:"bucket"`
+	// AWSRegion is the region the bucket is in, if it is an S3 bucket
+	AWSRegion string `json:"aws_region,omitempty"`
+	// AWSProfile is the AWS CLI config profile that should be used to
+	// authenticate to AWS, if the bucket is an S3 bucket
+	AWSProfile string `json:"aws_profile,omitempty"`
+}
+
+// NewStorage creates an instance of the appropriate implementation of Writer for
+// the provided bucket
+func NewStorage(bucket *Bucket) (Storage, error) {
+	if bucket.AWSRegion != "" {
```

Comment:
I'd recommend you make a specific Bucket `isAWS` and `isGCS` methods rather than this implicit "does AWS Region have data?" - you should let the helper method check AWSProfile as well to avoid mistakes in the future.

Reply:
I'll do you one better: I changed the definition of `Bucket` and the Terraform output so that instead of bucket name, we get a URL-like string in the form `gs://bucket-name` or `s3://bucket-name` so that `NewStorage` can more reliably determine which cloud storage service we're using. All other things being equal, I might have added an explicit field to `Bucket` to indicate which cloud platform we're using, but we use this bucket URL convention in several other places throughout `prio-server` so I'm going to err on the side of consistency.